src/main/java/com/github/distributedtracing/ContinuationLocal.java

```diff
@@ -0,0 +1,170 @@
+package com.github.distributedtracing;
+
+public final class ContinuationLocal<T> {
+    private static InheritableThreadLocal<Object[]> localCtx =
+            new InheritableThreadLocal<Object[]>();
+    private static volatile int size = 0;
+
+    /**
+     * Return a snapshot of the current ContinuationLocal state.
+     */
+    public static Object[] save() {
```

Comment:
It seems dropping some type safety and returning `Object` would be superior over `Object[]`. That 1) allows the concrete implementation to morph and 2) discourages modifying the object. It is true that users could cast to the `Object[]`, but that seems less of a concern. `restore()` would require "saved was returned from save()" and do a cast internally, throwing an `IllegalStateException` if it isn't the expected type.

That isn't to say that I think returning an `Object` is better than a special type needing an extra allocation, just that it seems less problematic than `Object[]`.